src/hpct-importer.1
.\" ========================================================================
.\"    _   _ _ ____            ____          _____
.\"   | | | (_)  _ \ ___ _ __ / ___|___  _ _|_   _| __ __ _  ___ ___ _ __
.\"   | |_| | | |_) / _ \ '__| |   / _ \| '_ \| || '__/ _` |/ __/ _ \ '__|
.\"   |  _  | |  __/  __/ |  | |__| (_) | | | | || | | (_| | (_|  __/ |
.\"   |_| |_|_|_|   \___|_|   \____\___/|_| |_|_||_|  \__,_|\___\___|_|
.\"
.\"      ---  High-Performance Connectivity Tracer (HiPerConTracer)  ---
.\"                https://www.nntb.no/~dreibh/hipercontracer/
.\" ========================================================================
.\"
.\" High-Performance Connectivity Tracer (HiPerConTracer)
.\" Copyright (C) 2015-2026 by Thomas Dreibholz
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.\" Contact: dreibh@simula.no
.\"
.\" ###### Setup ############################################################
.Dd September 7, 2026
.Dt hpct\-importer 1
.Os hpct\-importer
.\" ###### Name #############################################################
.Sh NAME
.Nm hpct\-importer
.Nd HiPerConTracer Trace Data Importer
.\" ###### Synopsis #########################################################
.Sh SYNOPSIS
.Nm hpct\-importer
.Op Fl C Ar importer\_\%configuration | Fl \-importer\-\%config Ar importer\_\%configuration
.br
.Op Fl D Ar database\_\%configuration | Fl \-database\-\%config Ar database\_\%configuration
.br
.Op Fl L Ar level | Fl \-loglevel Ar level
.br
.Op Fl Z Ar on|off | Fl \-logcolor Ar on|off
.br
.Op Fl O Ar file | Fl \-logfile Ar file
.br
.Op Fl q | Fl \-quiet
.Op Fl w | Fl \-verbose
.br
.Op Fl X Ar mode | Fl \-import\-mode Ar mode
.br
.Op Fl Y Ar depth | Fl \-import\-max\-depth Ar depth
.br
.Op Fl I Ar path | Fl \-import\-file\-path Ar path
.br
.Op Fl B Ar path | Fl \-bad\-file\-path Ar path
.br
.Op Fl G Ar path | Fl \-good\-file\-path Ar path
.br
.Op Fl F Ar filter\_\%regexp | Fl \-import\-\%file\-\%path\-\%filter Ar filter\_\%regexp
.br
.Op Fl Q | Fl \-quit\-when\-idle
.br
.Op Fl \-ping\-workers Ar number
.Op Fl \-ping\-files Ar number
.br
.Op Fl \-traceroute\-workers Ar number
.Op Fl \-traceroute\-files Ar number
.\" .br
.\" .Op Fl Fl jitter-workers Ar number
.\" .Op Fl Fl jitter-files Ar number
.Nm hpct\-importer
.Op Fl h | Fl \-help
.Nm hpct\-importer
.Op Fl v | Fl \-version
.\" ###### Description ######################################################
.Sh DESCRIPTION
.Nm hpct\-importer
imports HiPerConTracer results into an SQL or NoSQL database. Currently, the
following database backends are supported:
MySQL/MariaDB (SQL),
PostgreSQL (SQL),
MongoDB (NoSQL).
.br
Results files imported successfully can be moved into a "good" directory,
deleted or kept as\-in (for debugging only!).
Results files which failed to import are moved into a "bad" directory.
.br
The input of multiple input files can be combined into a single database
transaction, to improve performance on short files. In case of an import
error, each file is imported sequentially to identify faulty ones.
.Pp
.\" ###### Arguments ########################################################
.Sh ARGUMENTS
The following arguments may be provided:
.Bl -tag -width indent
.It Fl L Ar level | Fl \-loglevel Ar level
Sets the minimum logging level to the provided value: 0=trace, 1=debug, 2=info, 3=warning, 4=error, 5=fatal.
.It Fl Z Ar on|off | Fl \-logcolor Ar on|off
Enables (on) or disables (off) the usage of ANSI color escape sequences to colorise the log output according to its level.
.It Fl O Ar file | Fl \-logfile Ar file
Sets a file name for appending the log output to. If a file is set, no log output is made to the console.
Otherwise, all log output will be written to the console.
.It Fl q | Fl \-quiet
Sets the minimum logging level to 3 (warning).
.It Fl w | Fl \-verbose
Sets the minimum logging level to 0 (trace).
.It Fl C Ar importer\_configuration | Fl \-importer\-config Ar importer\_configuration
Sets the name of the importer configuration file. This option must be provided!
.br
See /usr/share/hipercontracer/hipercontracer\-importer.conf for an example.
.It Fl D Ar database\_configuration | Fl \-database\-config Ar database\_configuration
Sets the name of the database configuration file. This option must be provided!
.br
See /usr/share/hipercontracer/hipercontracer\-database.conf for an example.
.It Fl X Ar mode | Fl \-import\-mode Ar mode
Specifies the import mode,
overriding the mode read from the database configuration file:
DeleteImportedFiles (delete successfully imported files)
MoveImportedFiles (move successfully imported files into "good" directory),
KeepImportedFiles (keep successfully imported files where they are).
Do not use KeepImportedFiles except for testing. Otherwise, the next re\-run of
hpct\-importer tries to import them again! They may create duplicates when the
underlying database system (e.g. NoSQL without unique indices) does not ensure
uniqueness!
.It Fl Y Ar depth | Fl \-import\-max\-depth Ar depth
Specifies the import directory traversal depth,
overriding the mode read from the database configuration file.
.It Fl I Ar path | Fl \-import\-file\-path Ar path
Specifies the import file path,
overriding the mode read from the database configuration file.
This path is used to look for files to import.
.It Fl B Ar path | Fl \-bad\-file\-path Ar path
Specifies the "good" file path,
overriding the mode read from the database configuration file.
This path is used to move successfully imported files to, when
the import mode is set to KeepImportedFiles.
.It Fl G Ar path | Fl \-good\-file\-path Ar path
Specifies the "bad" file path,
overriding the mode read from the database configuration file.
This path is used to move files to which failed to import.
.It Fl F Ar filter\_regexp | Fl \-import\-file\-path\-filter Ar filter\_regexp
Specifies the import path filter,
overriding the mode read from the database configuration file.
This filter can be used to only import a subset of the import files for
testing. The filter expression is an std::regex regular expression.
.It Fl Q | Fl \-quit\-when\-idle
Quit importer when all workers are idle. This will perform only one import
round.
.It Fl \-ping\-workers Ar number
Sets the number of Ping import workers. Default: 1.
An import file is distributed to a worker by hashing the source address
given in the file name.
.It Fl \-ping\-files Ar number
Sets the number of Ping import files per database transaction. Default: 1.
.It Fl \-traceroute\-workers Ar number
Sets the number of Traceroute import workers. Default: 1.
An import file is distributed to a worker by hashing the source address
given in the file name.
.It Fl \-traceroute\-files Ar number
Sets the number of Traceroute import files per database transaction. Default: 1.
.\" .It Fl Fl jitter-workers Ar number
.\" Sets the number of Jitter import workers. Default: 1.
.\" An import file is distributed to a worker by hashing the source address
.\" given in the file name.
.\" .It Fl Fl jitter-files Ar number
.\" Sets the number of Jitter import files per database transaction. Default: 1.
.It Fl h | Fl \-help
Prints command help.
.It Fl v | Fl \-version
Prints program version.
.El
.\" ###### Exit status ######################################################
.Sh EXIT STATUS
The
.Nm
tool exits with 0 on success, and >0 in case of an error.
.\" ###### Examples #########################################################
.Sh EXAMPLES
.Bl -tag -width indent
.It hpct\-importer \-C hipercontracer\-importer.conf \-D hipercontracer\-database.conf
.It hpct\-importer \-\-importer\-config hipercontracer\-importer.conf \-\-database\-config hipercontracer\-database.conf \-\-verbose
.It hpct\-importer \-C hipercontracer\-importer.conf \-D hipercontracer\-database.conf \-F "1234/.*"
.It hpct\-importer \-\-version
.It hpct\-importer \-h
.El
.\" ###### See also #########################################################
.Sh SEE ALSO
.Xr dbshell 1
.Xr hipercontracer 1
.Xr hpct-query 1
.Xr hpct-results 1
.Xr hpct-viewer 1
.\" ###### Notes ############################################################
.Sh NOTES
This program is part of HiPerConTracer. The latest version of HiPerConTracer can be found on the HiPerConTracer Homepage at
.Lk https://www.nntb.no/\(tidreibh/hipercontracer/ "HiPerConTracer Homepage" .
.\" ###### Authors ##########################################################
.Sh AUTHORS
Thomas Dreibholz,
.Lk https://www.nntb.no/\(tidreibh "Homepage"


ChangeLog
commit c2586671eedee0068fc623622fe59f058b1e40c2
Merge: b00a9968 1c3000d6
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 11:59:49 2026 +0200

    Merge branch 'dreibh/homebrew' of github.com:dreibh/hipercontracer into dreibh/homebrew

commit b00a9968e3a706ba6c4b79e7fd74d51753cc3ba1
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 11:58:41 2026 +0200

    Removed exit() call from library. DatabaseConfiguration::readConfiguration() returns false in case of a path error.

commit 1c3000d647dae82789040f95ff5a1b8b3853a65c
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 11:55:16 2026 +0200

    Updated RPM spec file.

commit a2188f7125d634ff6e0d0ede00c527730536205e
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 11:47:00 2026 +0200

    Using CMAKE_INSTALL_SYSCONFDIR for configuration file installation instead of hard-coding /etc.

commit 52ac9bc09bb4165923d0e3b2781497cab9d9f3fb
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 11:30:42 2026 +0200

    Improved Homebrow path handling in CMakeLists.txt.

commit 2be6cc58880f261804db093b06920e4f13522ecb
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 11:10:01 2026 +0200

    Some progress with the Homebrew port.

commit 21a740847032ad7e2b86da2571211a37e641d2bc
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 22:57:47 2026 +0200

    Some improvements.

commit f8a3f894156a3938dc44c7e3015ee4b50ab2142d
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 22:43:15 2026 +0200

    Initial homebrew packaging. Not working, yet!

commit 112761ee05490e780af98b8b80b29d4667c3ac86
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 17:07:58 2026 +0200

    Improved RPM spec file.

commit 4ab4e15001f36985d99663396bf93af46bd22056
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 16:57:57 2026 +0200

    Improved documentation.

commit 2f71eef68655e24b53083325c2bf7b6c569f3cd0
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 16:38:42 2026 +0200

    Setting capabilities (setcap cap_net_raw=ep) for Debian/Ubuntu and Fedora package.

commit 669311cc09b610fab6a00f170830bd9535d05980
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 16:37:25 2026 +0200

    Linux: Allow HiPerConTracer to run with capabilities (setcap cap_net_raw=ep) instead of startup with root permissions.

commit ef1049418431815e5674099d26b8d55f69770031
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 16:00:14 2026 +0200

    Another set of manpage improvements.

commit 3763e8cea0b34be3a7d0b76bc70b7e320f27f2f9
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 15:54:43 2026 +0200

    Further manpage improvements.

commit b0f30cc37b8ad936594ea882a6a80d07f05d9a9c
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 15:37:42 2026 +0200

    Updated manpages.

commit f2ddd54b04e55f370bf96964937d7586b93abcad
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Sep 7 15:37:23 2026 +0200

    Use positional parameters as destinations.

commit b701a523bb3e3177283a1254527dfb63f121b216
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Sep 4 18:22:15 2026 +0200

    Updated APKBUILD file.

commit cc74111202cfbc62d05470f40a9f4fd29b1b0d1c
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Sep 4 18:07:12 2026 +0200

    New release hipercontracer-2.2.10.

commit e2318e7c008132ecf89cb3391c9906a92d3962a7
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Sep 4 15:10:01 2026 +0200

    Fix for Debian test 30-dbeaver-tools.

commit 813403b8077be2520460d15674cd634399ec11db
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 16:23:32 2026 +0200

    Updated APKBUILD file.

commit 5c3c467cccde4b0e0fb78868650085f2fd890c94
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 15:08:02 2026 +0200

    Improved example scripts.

commit bb7d3cbfe79dc06f3c4b2a5fc3687cbb2d5b1c80
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 14:57:44 2026 +0200

    Updated script.

commit 25c5684be31fec6f4c923e94461509b4c71552ae
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 14:54:06 2026 +0200

    Added test script.

commit bb15335db526f34f8322d27270d63064eed58e3f
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 14:51:51 2026 +0200

    Added Ping example script as well.

commit f2cdf7327f20c5328b9b075839ce913bd83a56de
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 14:46:11 2026 +0200

    Further clean-ups.

commit b29a86abd30967b176eb5266098e2e16039d24e4
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 14:39:31 2026 +0200

    Further improvements of the examples, and some function name clean-ups.

commit 14334a09f1948fa48c0648274b11b13839ef9da9
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 12:53:25 2026 +0200

    Added another Python example.

commit 3269229394f112f1196d532d39992d2e344201a3
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 12:53:17 2026 +0200

    Improved Python library.

commit 3444d978211dfbab34ca19ced5eec21c350dd822
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Aug 31 16:33:02 2026 +0200

    New release hipercontracer-2.2.9.

commit b9f4f55dab00295fc3c299777a4718b630170651
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Aug 31 16:29:16 2026 +0200

    Version bump.

commit 1847ee70e030bc1c5fff9c9a0256b83af48a0ad2
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Aug 31 15:28:13 2026 +0200

    Updated test scripts.

commit 269f102483fea833bfd2aae28e8b9036ab180e89
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Aug 31 14:51:17 2026 +0200

    Manpage fix.

commit cebc281513d0e5ee95bae5eec470356485305c1a
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Aug 31 14:49:31 2026 +0200

    Improved manpages.

commit c3cc031b9c8a3363ab2ddbc2af3a028163a1ee72
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Aug 11 15:43:18 2026 +0200


src/hpct-importer.cc
// ==========================================================================
//     _   _ _ ____            ____          _____
//    | | | (_)  _ \ ___ _ __ / ___|___  _ _|_   _| __ __ _  ___ ___ _ __
//    | |_| | | |_) / _ \ '__| |   / _ \| '_ \| || '__/ _` |/ __/ _ \ '__|
//    |  _  | |  __/  __/ |  | |__| (_) | | | | || | | (_| | (_|  __/ |
//    |_| |_|_|_|   \___|_|   \____\___/|_| |_|_||_|  \__,_|\___\___|_|
//
//       ---  High-Performance Connectivity Tracer (HiPerConTracer)  ---
//                 https://www.nntb.no/~dreibh/hipercontracer/
// ==========================================================================
//
// High-Performance Connectivity Tracer (HiPerConTracer)
// Copyright (C) 2015-2026 by Thomas Dreibholz
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
// Contact: dreibh@simula.no

#include "logger.h"

#include "package-version.h"
#include "reader-jitter.h"
#include "reader-ping.h"
#include "reader-traceroute.h"
#include "universal-importer.h"

#include <filesystem>
#include <iostream>

#include <boost/program_options.hpp>



// ###### Main program ######################################################
int main(int argc, char** argv)
{
   // ====== Initialize =====================================================
   unsigned int          logLevel;
   bool                  logColor;
   std::filesystem::path logFile;
   std::filesystem::path databaseConfigFile;
   std::filesystem::path importerConfigFile;
   std::string           importModeName;
   unsigned int          importMaxDepth;
   std::filesystem::path importFilePath;
   std::filesystem::path badFilePath;
   std::filesystem::path goodFilePath;
   std::string           importFilePathFilter;
   bool                  quitWhenIdle;
   unsigned int          pingWorkers;
   unsigned int          tracerouteWorkers;
   unsigned int          jitterWorkers;
   unsigned int          pingTransactionSize;
   unsigned int          tracerouteTransactionSize;
   unsigned int          jitterTransactionSize;

   boost::program_options::options_description commandLineOptions;
   commandLineOptions.add_options()
      ( "help,h",
           "Print help message" )
      ( "version,v",
           "Show program version" )

      ( "loglevel,L",
           boost::program_options::value<unsigned int>(&logLevel)->default_value(boost::log::trivial::severity_level::info),
           "Set logging level" )
      ( "logfile,O",
           boost::program_options::value<std::filesystem::path>(&logFile)->default_value(std::filesystem::path()),
           "Log file" )
      ( "logcolor,Z",
           boost::program_options::value<bool>(&logColor)->default_value(true),
           "Use ANSI color escape sequences for log output" )
      ( "verbose,w",
           boost::program_options::value<unsigned int>(&logLevel)->implicit_value(boost::log::trivial::severity_level::trace),
           "Verbose logging level" )
      ( "quiet,q",
           boost::program_options::value<unsigned int>(&logLevel)->implicit_value(boost::log::trivial::severity_level::warning),
           "Quiet logging level" )

      ( "importer-config,C",        boost::program_options::value<std::filesystem::path>(&importerConfigFile),      "Importer configuration file")
      ( "database-config,D",        boost::program_options::value<std::filesystem::path>(&databaseConfigFile),      "Database configuration file")

      ("import-mode,X",             boost::program_options::value<std::string>(&importModeName),                    "Override import mode")
      ("import-max-depth,M",        boost::program_options::value<unsigned int>(&importMaxDepth)->default_value(0), "Override import max depth")
      ("import-file-path,I",        boost::program_options::value<std::filesystem::path>(&importFilePath),          "Override path for input files")
      ("bad-file-path,B",           boost::program_options::value<std::filesystem::path>(&badFilePath),             "Override path for bad files")
      ("good-file-path,G",          boost::program_options::value<std::filesystem::path>(&goodFilePath),            "Override path for good files")
      ("import-file-path-filter,F", boost::program_options::value<std::string>(&importFilePathFilter),              "Override import path filter (regular expression)")
      ("quit-when-idle,Q",
          boost::program_options::value<bool>(&quitWhenIdle)->implicit_value(true)->default_value(false),
          "Quit importer when idle")

      ( "ping-workers",
           boost::program_options::value<unsigned int>(&pingWorkers)->default_value(1),
           "Number of Ping import worker threads" )
      ( "ping-files",
           boost::program_options::value<unsigned int>(&pingTransactionSize)->default_value(1),
           "Number of Ping files per transaction" )
      ( "traceroute-workers",
           boost::program_options::value<unsigned int>(&tracerouteWorkers)->default_value(1),
           "Number of Traceroute import worker threads" )
      ( "traceroute-files",
           boost::program_options::value<unsigned int>(&tracerouteTransactionSize)->default_value(1),
           "Number of Traceroute files per transaction" )
      ( "jitter-workers",
           boost::program_options::value<unsigned int>(&jitterWorkers)->default_value(1),
           "Number of Jitter import worker threads" )
      ( "jitter-files",
           boost::program_options::value<unsigned int>(&jitterTransactionSize)->default_value(1),
           "Number of Jitter files per transaction" )
   ;

   // ====== Handle command-line arguments ==================================
   boost::program_options::variables_map vm;
   try {
      boost::program_options::store(boost::program_options::command_line_parser(argc, argv).
                                       style(
                                          boost::program_options::command_line_style::style_t::unix_style
                                       ).
                                       options(commandLineOptions).
                                       run(), vm);
      boost::program_options::notify(vm);
   }
   catch(std::exception& e) {
      std::cerr << "ERROR: Bad parameter: " << e.what() << "\n";
      return 1;
   }

   if(vm.count("help")) {
       std::cerr << "Usage: " << argv[0] << " OPTIONS" << "\n"
                 << commandLineOptions;
       return 1;
   }
   else if(vm.count("version")) {
      std::cerr << "HPCT Importer" << " " << HPCT_VERSION << "\n";
      return 0;
   }

   if(importerConfigFile.empty()) {
      std::cerr << "ERROR: No importer configuration file provided!\n";
      return 1;
   }
   if(databaseConfigFile.empty()) {
      std::cerr << "ERROR: No database configuration file provided!\n";
      return 1;
   }
   if(pingWorkers + tracerouteWorkers < 1) {
      std::cerr << "ERROR: At least one worker is needed!\n";
      return 1;
   }

   // ====== Read Importer configuration ====================================
   ImporterConfiguration importerConfiguration;
   if(!importerConfiguration.readConfiguration(importerConfigFile)) {
      exit(1);
   }
   if(importModeName.size() > 0) {
      if(!importerConfiguration.setImportMode(importModeName)) return 1;
   }
   if(importMaxDepth) {
      if(!importerConfiguration.setImportMaxDepth(importMaxDepth)) return 1;
   }
   if(importFilePath.string().size() > 0) {
      if(!importerConfiguration.setImportFilePath(importFilePath)) return 1;
   }
   if(goodFilePath.string().size() > 0) {
      if(!importerConfiguration.setGoodFilePath(goodFilePath)) return 1;
   }
   if(goodFilePath.string().size() > 0) {
      if(!importerConfiguration.setGoodFilePath(goodFilePath)) return 1;
   }
   if(badFilePath.string().size() > 0) {
      if(!importerConfiguration.setBadFilePath(badFilePath)) return 1;
   }
   if(importFilePathFilter.size() > 0) {
      if(!importerConfiguration.setImportPathFilter(importFilePathFilter)) return 1;
   }
   if( (importerConfiguration.getImportMode() == ImportModeType::KeepImportedFiles) &&
       (!quitWhenIdle) ) {
      std::cerr << "ERROR: Import mode \"KeepImportedFiles\" is only useful with --quit-when-idle option!\n";
      exit(1);
   }

   // ====== Read database configuration ====================================
   DatabaseConfiguration databaseConfiguration;
   if(!databaseConfiguration.readConfiguration(databaseConfigFile)) {
      exit(1);
   }

   // ====== Initialise importer ============================================
   HPCT_LOG(info) << "Startup:\n" << importerConfiguration << databaseConfiguration;
   initialiseLogger(logLevel, logColor,
                    (!logFile.empty()) ? logFile.string().c_str() : nullptr);

   boost::asio::io_context ioContext;
   UniversalImporter importer(ioContext, importerConfiguration, databaseConfiguration);


   // ====== Initialise database clients and readers ========================
   // ------ HiPerConTracer Ping --------------------------
   DatabaseClientBase* pingDatabaseClients[pingWorkers];
   PingReader*         pingReader = nullptr;
   if(pingWorkers > 0) {
      for(unsigned int i = 0; i < pingWorkers; i++) {
         pingDatabaseClients[i] = databaseConfiguration.createClient();
         assert(pingDatabaseClients[i] != nullptr);
         if(!pingDatabaseClients[i]->open()) {
            exit(1);
         }
      }
      pingReader = new PingReader(importerConfiguration,
                                  pingWorkers, pingTransactionSize,
                                  importerConfiguration.getTableName(PingReader::Identification,
                                                                     PingReader::Identification));
      assert(pingReader != nullptr);
      importer.addReader(*pingReader,
                         (DatabaseClientBase**)&pingDatabaseClients, pingWorkers);
   }

   // ------ HiPerConTracer Traceroute --------------------
   DatabaseClientBase* tracerouteDatabaseClients[tracerouteWorkers];
   TracerouteReader*   tracerouteReader = nullptr;
   if(tracerouteWorkers > 0) {
      for(unsigned int i = 0; i < tracerouteWorkers; i++) {
         tracerouteDatabaseClients[i] = databaseConfiguration.createClient();
         assert(tracerouteDatabaseClients[i] != nullptr);
         if(!tracerouteDatabaseClients[i]->open()) {
            exit(1);
         }
      }
      tracerouteReader = new TracerouteReader(importerConfiguration,
                                              tracerouteWorkers, tracerouteTransactionSize,
                                              importerConfiguration.getTableName(TracerouteReader::Identification,
                                                                                 TracerouteReader::Identification));
      assert(tracerouteReader != nullptr);
      importer.addReader(*tracerouteReader,
                         (DatabaseClientBase**)&tracerouteDatabaseClients, tracerouteWorkers);
   }

   // ------ HiPerConTracer Jitter ------------------------
   DatabaseClientBase* jitterDatabaseClients[jitterWorkers];
   JitterReader*   jitterReader = nullptr;
   if(jitterWorkers > 0) {
      for(unsigned int i = 0; i < jitterWorkers; i++) {
         jitterDatabaseClients[i] = databaseConfiguration.createClient();
         assert(jitterDatabaseClients[i] != nullptr);
         if(!jitterDatabaseClients[i]->open()) {
            exit(1);
         }
      }
      jitterReader = new JitterReader(importerConfiguration,
                                      jitterWorkers, jitterTransactionSize,
                                      importerConfiguration.getTableName(JitterReader::Identification,
                                                                         JitterReader::Identification));
      assert(jitterReader != nullptr);
      importer.addReader(*jitterReader,
                         (DatabaseClientBase**)&jitterDatabaseClients, jitterWorkers);
   }


   // ====== Main loop ======================================================
   if(importer.start(quitWhenIdle) == false) {
      exit(1);
   }
   if(quitWhenIdle) {
     importer.waitForFinish();
   }
   ioContext.run();
   importer.stop();


   // ====== Clean up =======================================================
   if(jitterWorkers > 0) {
      delete jitterReader;
      jitterReader = nullptr;
      for(unsigned int i = 0; i < jitterWorkers; i++) {
         delete jitterDatabaseClients[i];
         jitterDatabaseClients[i] = nullptr;
      }
   }
   if(tracerouteWorkers > 0) {
      delete tracerouteReader;
      tracerouteReader = nullptr;
      for(unsigned int i = 0; i < tracerouteWorkers; i++) {
         delete tracerouteDatabaseClients[i];
         tracerouteDatabaseClients[i] = nullptr;
      }
   }
   if(pingWorkers > 0) {
      delete pingReader;
      pingReader = nullptr;
      for(unsigned int i = 0; i < pingWorkers; i++) {
         delete pingDatabaseClients[i];
         pingDatabaseClients[i] = nullptr;
      }
   }
   return 0;
}
